From 63e89cd04e602db3b69c837c2f1c60550cd2490e Mon Sep 17 00:00:00 2001
From: Zheng Zucheng <zhengzucheng@huawei.com>
Date: Tue, 10 May 2022 21:37:17 +0800
Subject: [PATCH 2071/2592] [Backport] kernel/cputime: do not update cputime
 when cpu offline

hulk inclusion
category: bugfix
bugzilla: https://bugzilla.openeuler.org/show_bug.cgi?id=37

Reference: https://gitee.com/openeuler/kernel/commit/1a66de656562c7ff063eeab81f3e459101508031

-------------------------------

If we enable 'use_sched_idle_time' at commend line, when
possile cpus > online cpus, we get the uncorrected idle
time. Return the pervious sum_idle_time to avoid this issue.

conflicts:
	kernel/sched/cputime.c

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Tested-by: Yu Xiangyang <yuxiangyang4@huawei.com>
Reviewed-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 kernel/sched/cputime.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/sched/cputime.c b/kernel/sched/cputime.c
index 44a6e41bc3aa..4848c3cdaff6 100644
--- a/kernel/sched/cputime.c
+++ b/kernel/sched/cputime.c
@@ -1161,6 +1161,10 @@ unsigned long long sched_get_idle_time(int cpu)
 	struct rq_cputime *rt = &per_cpu(rq_cputimes, cpu);
 	u64 hi = 0, si = 0;
 
+	/* Do not update cpu time when cpu offline */
+	if (!cpu_online(cpu))
+		return rt->sum_idle_time;
+
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
 	if (sched_clock_irqtime) {
 		hi = kcpustat_cpu(cpu).cpustat[CPUTIME_IRQ_IDLE];
-- 
2.17.1

